### PR TITLE
feat(carousel): add event on carousel slide with direction info

### PR DESCRIPTION
**Pull request by giampierobono**
Before submitting a pull request, please make sure you have at least performed the following:

 - [x] read and followed the [CONTRIBUTING.md](https://github.com/ng-bootstrap/ng-bootstrap/blob/master/CONTRIBUTING.md) guide.
 - [x] built and tested the changes locally.
 - [x] added/updated any applicable tests.
 - [x] added/updated any applicable API documentation.
 - [ ] added/updated any applicable demos.

### CODE REVIEW

**Comment by giampierobono** [comment]
Don't understand why the build fails...in local it works

File: src/carousel/carousel.ts [hunk 1]
```diff
@@ -196,6 +212,48 @@ export class NgbCarousel implements AfterContentChecked,
...
+   */
+  direction: NgbSlideCarouselEventDirection;
+}
+
+/**
+ * Enum to define the carousel slide event direction
+ */
+export enum NgbSlideCarouselEventDirection {
+  LEFT = <any>'left',
+  RIGHT = <any>'right'
```
Comment (wesleycho) [inline src/carousel/carousel.ts]: Why type these to `any` and not make use of the implicit string typing?

**Comment by wesleycho** [comment]
The build failure is some flakiness with Edge testing in Sauce Labs - that is to be looked into.

This PR looks mostly good to me.

File: src/carousel/carousel.ts (hunk 1 above)
Comment (giampierobono) [inline src/carousel/carousel.ts]: As far as I know, you cannot assign directly a string to an enum value. In fact if I remove <any>, typescript says: type string is no assignable to type NgbSlideCarouselEventDirection.
Is there another way to do it?

Comment (wesleycho) [inline src/carousel/carousel.ts]: Looks like you should just be able to do
```ts
export enum NgbSlideCarouselEventDirection {
  left,
  right
}
```
and use it like `NgbSlideCarouselEventDirection.left` to represent the string value `left`

File: src/carousel/carousel.ts [hunk 2]
```diff
@@ -141,6 +149,14 @@ export class NgbCarousel implements AfterContentChecked,
   cycleToSelected(slideIdx: string) {
     let selectedSlide = this._getSlideById(slideIdx);
     if (selectedSlide) {
+      if (selectedSlide.id !== this.activeId) {
+        this.onCarouselSlide.emit({
+          type: 'slide',
+          prevSlideId: this.activeId,
+          newSlideId: selectedSlide.id,
```
Comment (wesleycho) [inline src/carousel/carousel.ts]: Maybe these can be shortened to `prev` and `current`.

File: src/carousel/carousel.ts (hunk 1 above)
Comment (giampierobono) [inline src/carousel/carousel.ts]: but doing like this NgbSlideCarouselEventDirection.left and NgbSlideCarouselEventDirection.right values won't be 0 and 1 instead of 'left' and 'right'?

File: src/carousel/carousel.ts (hunk 2 above)
Comment (giampierobono) [inline src/carousel/carousel.ts]: Totally agree. I will change them. Thanks for the suggestion :)

File: src/carousel/carousel.ts [hunk 3]
```diff
@@ -196,6 +212,48 @@ export class NgbCarousel implements AfterContentChecked,
     return isFirstSlide ? (this.wrap ? slideArr[slideArr.length - 1].id : slideArr[0].id) :
                           slideArr[currentSlideIdx - 1].id;
   }
+
+  private _getSlideEventDirection(currentActiveSlideId: string, nextActiveSlideId: string):
+      NgbSlideCarouselEventDirection {
+    const currentActiveSlideIdx = this._getSlideIdxById(currentActiveSlideId);
+    const nextActiveSlideIdx = this._getSlideIdxById(nextActiveSlideId);
+
+    return currentActiveSlideIdx > nextActiveSlideIdx ? NgbSlideCarouselEventDirection.RIGHT :
```
Comment (mdoing) [inline src/carousel/carousel.ts]: I would expect that when you are on the first slide and click the back arrow that the direction would be RIGHT. However, when you are on the first slide with this logic it will always return LEFT, regardless of which direction you are moving in. The same thing applies if you are on the last slide, the direction will always be RIGHT, regardless of which direction you move in.

Comment (giampierobono) [inline src/carousel/carousel.ts]: You are totally right! Thanks a lot! 

I have just updated my PR. Now the logic is simpler, basically when you click on prev is always RIGHT and when you click on next is always LEFT. The computation based on the index is now done only clicking on the indicators.

File: src/carousel/carousel.ts [hunk 4]
```diff
@@ -43,7 +45,7 @@ export class NgbSlide {
   },
   template: `
     <ol class="carousel-indicators">
-      <li *ngFor="let slide of slides" [id]="slide.id" [class.active]="slide.id === activeId" (click)="cycleToSelected(slide.id)"></li>
+      <li *ngFor="let slide of slides" [id]="slide.id" [class.active]="slide.id === activeId" (click)="select(slide.id)"></li>
```
Comment (pkozlowski-opensource) [inline src/carousel/carousel.ts]: This changes the behaviour in subtle way: now we would reset timer on click. Was it intentional? If so we should add a test for it.

File: src/carousel/carousel.ts [hunk 5]
```diff
@@ -85,6 +87,12 @@ export class NgbCarousel implements AfterContentChecked,
    */
   @Input() activeId: string;
 
+  /**
+   * A carousel slide event fired when the slide transition is completed.
+   * See NgbSlideCarouselEvent for payload details
+   */
+  @Output() onCarouselSlide = new EventEmitter<NgbSlideCarouselEvent>();
```
Comment (pkozlowski-opensource) [inline src/carousel/carousel.ts]: can we name it just `slide`? We are in the carousel context so no need to add the `Carousel` in the name. Similarly we know that this is an event so no need for `on` (we are not adding it for other directives / events).

File: package.json [hunk 6]
```diff
@@ -27,8 +27,8 @@
   },
   "homepage": "https://github.com/ng-bootstrap/ng-bootstrap#readme",
   "dependencies": {
-    "@angular/core": "2.3.1",
     "@angular/common": "2.3.1",
+    "@angular/core": "2.3.1",
```
Comment (pkozlowski-opensource) [inline package.json]: Why this change? While it doesn't hurt it always best idea to avoid unrelated changes in a single PR.

File: demo/src/app/components/carousel/carousel.component.ts [hunk 7]
```diff
@@ -8,6 +8,7 @@ import {DEMO_SNIPPETS} from './demos';
       <ngbd-api-docs directive="NgbCarousel"></ngbd-api-docs>
       <ngbd-api-docs directive="NgbSlide"></ngbd-api-docs>
       <ngbd-api-docs-config type="NgbCarouselConfig"></ngbd-api-docs-config>
+      <ngbd-api-docs-class type="NgbSlideCarouselEvent"></ngbd-api-docs-class>
```
Comment (pkozlowski-opensource) [inline demo/src/app/components/carousel/carousel.component.ts]: I don't think this new functionality warrants a separate demo - 2 carousel side by side already look wired on our demo page so I would prefer adding more demos unless those really add a lot of value.

**Comment by pkozlowski-opensource** [comment]
@giampierobono thnx, it looks good generally speaking but I've left some comments that I would love to see addressed before we merge this one. Thnx again!

File: demo/src/app/components/carousel/carousel.component.ts (hunk 7 above)
Comment (giampierobono) [inline demo/src/app/components/carousel/carousel.component.ts]: Isn't it the way to update the documentation adding specific information about this new event?

Comment (pkozlowski-opensource) [inline demo/src/app/components/carousel/carousel.component.ts]: Documentation is generated automatically from the source code - as long as you've got JS doc on the `@Output` it will make it to the demo site.

File: src/carousel/carousel.ts (hunk 1 above)
Comment (thbt) [inline src/carousel/carousel.ts]: What about using an union type for the value (`'left' || 'right'`) instead of an enum ?

**Comment by pkozlowski-opensource** [comment]
@giampierobono are you planning to do further work on this PR? If so could you please remove the demo part and address all the remaining comments so I could merge this PR? Thnx!

**Comment by giampierobono** [comment]
@pkozlowski-opensource I have just reverted the demo changes and I have already applied all the other changes in comments. I was waiting for your reply to my comment about the documentation. 

The PR should be ok now. Let me know if something is still missing.

Thanks again.

File: src/carousel/carousel.ts [hunk 8]
```diff
@@ -196,6 +208,48 @@ export class NgbCarousel implements AfterContentChecked,
...
+}
+
+/**
+* The payload of the slide event fired when the slide transition is completed
+*/
+export interface NgbSlideCarouselEvent {
+  /**
+   * Event type
+   */
+  type: 'slide';
```
Comment (pkozlowski-opensource) [inline src/carousel/carousel.ts]: Why do we need `type`? Do we have other types and we need to distinguish between those? Isn't interface enough?

Comment (giampierobono) [inline src/carousel/carousel.ts]: When I have done this the first time, I have checked the bootstrap carousel doc and there are 2 events: slide and slid (the first is for transition start and the second for transition end if I remember correctly) and the initial plan was to reproduce the same behavior, that's why we have type. 
Need to add the slid event but I had not time till now, so I have pushed this. I can remove it if you want, I agree that since by now there is only one event it's not necessary.

Comment (pkozlowski-opensource) [inline src/carousel/carousel.ts]: Yeh, let's remove it. We should rather have 2 different event interfaces rather than trying to discriminate those by a string.

Comment (giampierobono) [inline src/carousel/carousel.ts]: of course 2 different events and not only one discriminated by strings :) it's the bs event structure having the 'type' prop so I have reproduced it. I am going to update the PR

**Comment by pkozlowski-opensource** [comment]
@giampierobono I was about to merge this PR but then I had a closer look and started to ask myself the following question: what is the use-case for exposing slide direction in the new event? Do you have a specific need? AFAIK it wasn't explicitly asked for in any of the prior issues (#1331, #1356).

If you don't have specific use-case I would prefer that we leave off this functionality as it adds a bit of code and currently people can't even visually see slide direction as we don't support animations (!)

**Comment by giampierobono** [comment]
@pkozlowski-opensource sorry I have just seen your message.

For me one use-case is keep in sync the carousel and some panels below it (with text adding more details to each carousel image), that will be displayed or not, following the event direction without relying only on the image id. Or another use case could be bind to the event and perform an action only if the direction is "right" or "left" and not both.

**Comment by pkozlowski-opensource** [comment]
@giampierobono sure, I can also _imagine_ use cases - I was rather asking if you, or any users that you know, already _has_ a specific use-case :-)

**Comment by giampierobono** [comment]
For something I am working on, the second use-case is the one I would need.

**Comment by kesarion** [comment]
Is something still blocking this pull from being merged?

**Comment by jspizziri** [comment]
@pkozlowski-opensource,

Is there further work on this PR before it get's merged in? I have a couple of use-cases that this would be greatly helpful for.

Thanks!
